ledger/participant-integration-api/src/test/lib/scala/platform/store/backend/StorageBackendTestsIngestion.scala

```diff
@@ -88,4 +93,46 @@ private[backend] trait StorageBackendTestsIngestion
     partiesAfterLedgerEndUpdate should not be empty
   }
 
+  private val NumberOfUpsertPackagesTests = 100
+  it should s"safely upsert packages concurrently ($NumberOfUpsertPackagesTests)" in withConnections(
+    2
+  ) { connections =>
+    import scala.concurrent.ExecutionContext.Implicits.global
+
+    def test() = {
+      val List(connection1, connection2) = connections
+      def packageFor(n: Int): DbDto.Package =
+        dtoPackage(offset(n.toLong))
+          .copy(package_id = s"abc123$n")
+      val conflictingPackageDtos = 11 to 20 map packageFor
+      val packages1 = 21 to 30 map packageFor
+      val packages2 = 31 to 40 map packageFor
+
+      executeSql(backend.parameter.initializeParameters(someIdentityParams))
+
+      def ingestPackagesF(connection: Connection, packages: Iterable[DbDto.Package]): Future[Unit] =
+        Future {
+          connection.setAutoCommit(false)
+          ingest(packages.toVector, connection)
+          connection.commit()
+        }
+
+      val ingestF1 = ingestPackagesF(connection1, packages1 ++ conflictingPackageDtos)
+      val ingestF2 = ingestPackagesF(connection2, packages2 ++ conflictingPackageDtos)
+
+      Await.result(ingestF1, Duration(10, "seconds"))
```

Comment:
Importing `DurationInt` allows slightly nicer syntax:
```suggestion
      Await.result(ingestF1, 10.seconds)
```